protocol/syncer.go

```diff
@@ -1,267 +1,146 @@
+				if receivedPayloads[msg.PayloadPosition] {
+					lgr.Trace("already processed this payload", "payload_position", msg.PayloadPosition, "peer_id", peerID)
 					continue
 				}
-				if receivedSectors[msg.SectorID] {
-					lgr.Trace("already processed this sector", "sector_id", msg.SectorID, "peer_id", peerID)
+				// TODO: if payloadposition = 0xff, handle equivocation proof
+				if opts.SectorSize != msg.PayloadPosition {
+					lgr.Trace("received unexpected payload position", "payload_size", opts.SectorSize, "payload_position", msg.PayloadPosition)
 					continue
 				}
-				if awaitingSectorID != int(msg.SectorID) {
-					lgr.Trace("received unsolicited sector", "sector_id", msg.SectorID, "peer_id", peerID)
-					continue
+				var sectorTipHash crypto.Hash = opts.PrevHash
+				for i := 0; int(i) < len(msg.Payload); i++ {
+					sectorTipHash = blob.SerialHashSector(msg.Payload[i], sectorTipHash)
 				}
-				hash := awaitingSectorHash(msg.SectorID, blob.HashSector(msg.Sector))
-				if expHash != hash {
-					lgr.Warn("invalid sector received", "sector_id", msg.SectorID, "peer_id", peerID)
+				if sectorTipHash != opts.SectorTipHash {
+					lgr.Trace("payload tip hash mismatch", "payload_tip_hash", sectorTipHash, "expected_payload_tip_hash", opts.SectorTipHash)
+					peer, err := opts.Mux.PeerByID(peerID)
+					if err != nil {
+						lgr.Trace("error fetching peer", "peer_id", peerID)
+					}
+					// TODO: set header.bannedat = time.now for this name
+					if err := peer.Close(); err != nil {
+						lgr.Trace("error banning peer", "peer_id", peerID)
+					}
+					// TODO: generate equivocation proof
 					continue
 				}
-				if err := opts.Tx.WriteSector(msg.SectorID, msg.Sector); err != nil {
-					lgr.Error("failed to write sector", "sector_id", msg.SectorID, "err", err)
-					continue
+				for i := 0; int(i) < len(msg.Payload); i++ {
+					if err := opts.Tx.WriteSector(msg.Payload[i]); err != nil {
+						lgr.Error("failed to write payload", "payload_id", i, "err", err)
+						continue
+					}
 				}
-				receivedSectors[msg.SectorID] = true
-				lgr.Debug(
-					"synced sector",
-					"name", opts.Name,
-					"sector_id", msg.SectorID,
-					"peer_id", peerID,
-				)
-				awaitingSectorID = -1
-				sectorProcessedCh <- struct{}{}
+				receivedPayloads[msg.PayloadPosition] = true
+				payloadProcessedCh <- struct{}{}
 			case <-doneCh:
 				return
 			}
 		}
 	}()
 
-sectorLoop:
-	for id := range reqdSectors {
-		timeout := time.NewTimer(opts.Timeout)
-		lgr.Debug("requesting sector", "id", id)
-		sectorReqCh <- id
+payloadLoop:
+	for {
+		lgr.Debug("requesting payload")
 		select {
-		case <-sectorProcessedCh:
-			lgr.Debug("sector processed", "id", id)
-			delete(outReqdSectors, id)
-		case <-timeout.C:
-			lgr.Warn("sector request timed out", "id", id)
-			break sectorLoop
+		case <-payloadProcessedCh:
+			lgr.Debug("payload processed")
+			break payloadLoop
+		case <-time.NewTimer(opts.Timeout).C:
```

Comment:
Timeout is being redefined with every iteration of the for loop - is this expected?

Reply:
Moved it out of the loop.